src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+                handles.push(handle);
+            }
+
+            // On the main thread...
+
+            // Drop the main thread's handle to the input receiver because, while we're still
+            // sending data, the way the main thread can learn that a worker thread has already
+            // errored and so we should stop sending data is to get a `SendError` from input
+            // channel when trying to send additional data to the worker threads, which will only
+            // happen if our handle to the receiver has been dropped.
+            drop(rx_input);
+
+            let buf_capacity = usize::next_power_of_two(NBYTES_PER_STACK_GUESS * nstacks_per_job);
+            let mut buf = Vec::with_capacity(buf_capacity);
+            let (mut index, mut nstacks) = (0, 0);
+
+            loop {
+                let n = reader.read_until(b'\n', &mut buf)?;
+                if n == 0 {
+                    // If we've reached the end of the data, send the final chunk to the worker
+                    // threads and break from the loop, The worker threads may or may not still
+                    // be alive (depending on if one errored in between the sending of the last
+                    // chunk and the sending of this one), but either way we should break the loop;
+                    // so there's no need to check for a `SendError` here.
+                    let _ = tx_input.send(Some(buf));
+                    break;
+                }
+                let line = &buf[index..index + n];
+                index += n;
+                if self.is_end_of_stack(line) {
+                    // If we've reached the end of a stack, count it.
+                    nstacks += 1;
+                    if nstacks == nstacks_per_job {
+                        // If we've accumulated enough stacks to make up a chunk to send to the
+                        // worker threads, try to send it.
+                        let buf_capacity = usize::next_power_of_two(buf.capacity());
+                        let chunk = mem::replace(&mut buf, Vec::with_capacity(buf_capacity));
+                        if tx_input.send(Some(chunk)).is_err() {
+                            // If sending the chunk produces a `SendError`, this means that one
+                            // of the worker threads has errored, sent a signal to all the other
+                            // worker threads to shut down, and they have all shutdown, in which
+                            // case we know there will be an error waiting for us on the error
+                            // channel; so we should stop parsing input data (i.e. break).
+                            break;
+                        }
+                        index = 0;
+                        nstacks = 0;
+                    }
+                    continue
+                }
+            }
+
+            // We've run out of input data to send to the worker threads; so tell them.
+            for _ in &handles {
```

Comment:
With the removal of the `Option` I sketched above, I think this whole loop can just become `drop(handles)`.